jets/e/secp.c

```diff
@@ -0,0 +1,356 @@
+{
+  // build library context object once (and only once)
+  static secp256k1_context * context = NULL;
+  if (NULL == context){
+    context = secp256k1_context_create(SECP256K1_CONTEXT_SIGN);
+  } 
+
+  // parse arguments, convert endianness
+  c3_y has_be[32];
+  c3_y prv_be[32];
+  u3r_bytes_reverse(0, 32, has_be, has);
+  u3r_bytes_reverse(0, 32, prv_be, prv);
+
+  
+  // sign
+  //    N.B. if we want the 'v' field we can't use default secp256k1_ecdsa_sign(), 
+  //    but must use secp256k1_ecdsa_sign_recoverable()
+  c3_ws ret;
+  secp256k1_ecdsa_recoverable_signature sig;
+  ret = secp256k1_ecdsa_sign_recoverable(context,                           // IN: context object
+                                         & sig,                             // OUT: signature
+                                         (const c3_y *) has_be,  // IN: 32 byte hash to be signed
+                                         (const c3_y *) prv_be,  // IN: 32 byte secret key
+                                         (secp256k1_nonce_function) NULL,   // IN: nonce-function ptr ; NULL = default
+                                         (const void *) NULL                // IN: data for nonce function; not used 
+                                         );
+  if (1 != ret) {
+    fprintf(stderr, "\rsecp jet: crypto package error\n");
+    return u3m_bail(c3__exit);
+  }
+
+  // convert opaque 65 byte signature into v + [r + s]
+  // convert endianness while we're at it
+  c3_y output64[64];  
+  c3_ws v = 0;
+  ret = secp256k1_ecdsa_recoverable_signature_serialize_compact(context,
+                                                                output64, // OUT: 64 byte sig (R,S)
+                                                                & v,      // OUT: V
+                                                                & sig);   // IN:  65 byte sig
+  if (1 != ret) {
+    fprintf(stderr, "\rsecp jet: crypto package error\n");
+    return u3m_bail(c3__exit);
+  }
+
+  c3_y s[32];
+  c3_y r[32];
+  byte_reverse(output64,      r, 32);
+  byte_reverse(output64 + 32, s, 32);
+
+  // package s,r,v signature for return
+  v += 27;
+  u3_noun s_noun  = u3i_words(8, (const c3_w*) s);
+  u3_noun r_noun  = u3i_words(8, (const c3_w*) r);
+  return (u3nt(v, r_noun, s_noun));
+
+}
+
+
+// recover pubkey from signature (which is how we verify signatures) 
+//
+
+u3_noun
+u3we_reco(u3_noun cor)
+{
+  u3_noun has, sig_v, sig_r, sig_s;
+  
+  if ( (c3n == u3r_mean(cor,
+                        u3x_sam_2,  &has,
+                        u3x_sam_6,  &sig_v,
+                        u3x_sam_14,  &sig_r,
+                        u3x_sam_15,  &sig_s,
+                        0)) ||
+       (c3n == u3ud(has)) ||
+       (c3n == u3ud(sig_v)) ||
+       (c3n == u3ud(sig_r)) ||
+       (c3n == u3ud(sig_s)) )
+    {
+      fprintf(stderr, "\rsecp jet: crypto package error\n");
+      return u3m_bail(c3__exit);
+    } else {
+    return u3qe_reco(has, sig_v, sig_r, sig_s);
+  }
+
+
```

Comment:
Some of the functions have leading and/or trailing whitelines. We should be able to fit the braces snugly around the function code. And double (or triple) whitelines in general can be replaced with just one.

Reply:
done